workspaces/mcp-integrations/plugins/software-catalog-mcp-tool/src/plugin.ts

```diff
@@ -0,0 +1,334 @@
+                            'The owner of the Backstage entity (e.g., team-platform, user:john.doe)',
+                          ),
+                        lifecycle: z
+                          .string()
+                          .optional()
+                          .describe(
+                            'The lifecycle of the Backstage entity (e.g., production, staging, development)',
+                          ),
+                        dependsOn: z
+                          .array(z.string())
+                          .optional()
+                          .describe(
+                            'List of entities this entity depends on (e.g., component:default/database)',
+                          ),
+                      }),
+                      z.custom<Entity>(),
+                    ]),
+                  )
+                  .describe(
+                    'An array of entities (either Backstage Entity objects or shortened entity information based on verbose parameter)',
+                  ),
+                error: z
+                  .string()
+                  .optional()
+                  .describe('Error message if validation fails'),
+              }),
+          },
+          action: async ({ input }) => {
+            // Validate that type is only used with kind -- we could just allow `type` to be specified without `kind` but given types are per kind it made sense to restrict it
+            // The Backstage MCP server will return a 500 error if we throw a validation error (without saying why), so instead, let's return the error message in the output
+            // TODO: Investigate potential upstream improvements to allow error messages to be returned to the client
+            if (input.type && !input.kind) {
+              return {
+                output: {
+                  entities: [],
+                  error:
+                    'entity type cannot be specified without an entity kind specified',
+                },
+              };
+            }
+            try {
+              const result = await fetchCatalogEntities(
+                catalog,
+                auth,
+                logger,
+                input,
+              );
+              return {
+                output: {
+                  ...result,
+                  error: undefined,
+                },
+              };
+            } catch (error) {
+              logger.error(
+                'fetch-catalog-entities: Error fetching catalog entities:',
+                error,
+              );
+              return {
+                output: {
+                  entities: [],
+                  error: error.message,
+                },
+              };
+            }
+          },
+        });
+      },
+    });
+  },
+});
+
+// fetchCatalogEntities retrieves the list of entities present in the Backstage catalog, with optional filtering by kind and type
+export async function fetchCatalogEntities(
```

Comment:
hopefully once we are all published in NPM we can start building common libs for stuff like this